packages/sdk/src/adapters/multipeer/clientDesyncPreprocessor.ts

```diff
@@ -0,0 +1,47 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Client, MissingRule, Rules } from '.';
+import { Message } from '../..';
+import { Middleware } from '../../protocols';
+import { UserJoined } from '../../types/network/payloads';
+import { ExportedPromise } from '../../utils/exportedPromise';
+
+/**
+ * Filter user-exclusive actors to a queue, then flush them after user-join
+ * @hidden
+ */
+export class ClientDesyncPreprocessor implements Middleware {
+    constructor(private client: Client) { }
+    /** @hidden */
+    public beforeSend(message: Message, promise?: ExportedPromise): Message {
+        const payloadType = message.payload.type;
+        const rule = Rules[payloadType] || MissingRule;
+        const forUser = rule.client.shouldSendToUser(
+            message, this.client.userId, this.client.session, this.client);
+        if (forUser !== null && !this.client.userId) {
+            // this message is user-exclusive, and the client's user ID is not yet settled,
+            // queue and cancel send for now
+            this.client.userExclusiveMessages.push({ message, promise });
+        } else if (forUser === null || forUser === true && !!this.client.userId) {
+            // this message is intended for this client's user, send now
+            return message;
+        }
+        // this message is intended for a different user, discard
+    }
+
+    /** @hidden */
+    public beforeRecv(message: Message): Message {
+        if (message.payload.type === 'user-joined') {
+            const userJoin = message.payload as UserJoined;
+            this.client.userId = userJoin.user.id;
+            while (this.client.userExclusiveMessages.length > 0) {
+                const queuedMsg = this.client.userExclusiveMessages.splice(0, 1)[0];
```

Comment:
>this.client.userExclusiveMessages.splice(0, 1)[0] [](start = 34, length = 49)

This could be simplified to `this.client.userExclusiveMessages.splice(0);`

Reply:
`splice(0)` would remove all elements from the array, and return a new array with those contents. The way this is structured, I just want to pop the first element from the array. I could structure it differently I suppose.

---
In reply to: [276690006](https://github.com/Microsoft/mixed-reality-extension-sdk/pull/314#discussion_r276690006) [](ancestors = 276690006)